File: java-simple-http-server/JavaSimpleHttpServer/src/main/java/RequestHandler.java
import enums.Status;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;

public class RequestHandler {

    private static Path BAD_REQUEST_HTML_PATH = Paths.get("public/400.html");
    private static Path FORBIDDEN_HTML_PATH = Paths.get("public/403.html");
    private static Path NOT_FOUND_HTML_PATH = Paths.get("public/404.html");
    private static String HTML_MIME = "text/html;charset=utf8";
    private static MimeDetector mimeDetector = new MimeDetector("mimes.properties");

    public Response handlerRequest(Request request) throws IOException {
        if (request == null) {
            return new Response(Status.BAD_REQUEST, HTML_MIME, Files.readAllBytes(BAD_REQUEST_HTML_PATH));
        }

        Path resourcePath = Paths.get("public", request.path).normalize();

        if (!resourcePath.startsWith("public/")) {
            return new Response(Status.FORBIDDEN, HTML_MIME, Files.readAllBytes(FORBIDDEN_HTML_PATH));
        }

        if (Files.isRegularFile(resourcePath)) {
            String mime = mimeDetector.getMime(resourcePath);
            return new Response(Status.OK, mime, Files.readAllBytes(resourcePath));
        }

        Path indexFilePath = resourcePath.resolve("index.html");
        if (Files.isDirectory(resourcePath) && Files.exists(indexFilePath)) {
            return new Response(Status.OK, HTML_MIME, Files.readAllBytes(indexFilePath));
        }

        return new Response(Status.NOT_FOUND, HTML_MIME, Files.readAllBytes(NOT_FOUND_HTML_PATH));
    }
}

File: kotlin-simple-http-server/src/main/kotlin/RequestHandler.kt
import enums.Status
import java.io.IOException
import java.nio.file.Files
import java.nio.file.Paths

class RequestHandler {

    @Throws(IOException::class)
    fun handlerRequest(request: Request?): Response {
        if (request == null) {
            return Response(Status.BAD_REQUEST, HTML_MIME, Files.readAllBytes(BAD_REQUEST_HTML_PATH))
        }

        val resourcePath = Paths.get("public", request.path).normalize()

        println(resourcePath)

        if (!resourcePath.startsWith("public/")) {
            return Response(Status.FORBIDDEN, HTML_MIME, Files.readAllBytes(FORBIDDEN_HTML_PATH))
        }

        if (Files.isRegularFile(resourcePath)) {
            val mime = mimeDetector.getMime(resourcePath)
            return Response(Status.OK, mime, Files.readAllBytes(resourcePath))
        }

        val indexFilePath = resourcePath.resolve("index.html")

        return if (Files.isDirectory(resourcePath) && Files.exists(indexFilePath)) {
            Response(Status.OK, HTML_MIME, Files.readAllBytes(indexFilePath))
        } else Response(Status.NOT_FOUND, HTML_MIME, Files.readAllBytes(NOT_FOUND_HTML_PATH))

    }

    companion object {

        private val BAD_REQUEST_HTML_PATH = Paths.get("public/400.html")
        private val FORBIDDEN_HTML_PATH = Paths.get("public/403.html")
        private val NOT_FOUND_HTML_PATH = Paths.get("public/404.html")
        private const val HTML_MIME = "text/html;charset=utf8"
        private val mimeDetector = MimeDetector("mimes.properties")
    }
}
